murenn/dtcwt/nn.py

```diff
@@ -38,38 +43,64 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
                 groups=in_channels,
                 padding="same",
             )
-            torch.nn.init.normal_(conv1d_j.weight)
+            torch.nn.init.normal_(conv1d_j.weight, std=1/math.sqrt(T))
             conv1d.append(conv1d_j)
+
         self.down = torch.nn.ModuleList(down)
         self.conv1d = torch.nn.ParameterList(conv1d)
 
 
     def forward(self, x):
         """
         Args:
-            x (PyTorch tensor): Input data. Should be a tensor of shape
-                `(B, C, T)` where B is the batch size, C is the number of
-                channels and T is the number of time samples.
-                Note that T must be a multiple of 2**J, where J is the number
-                of wavelet scales (see documentation of MuReNNDirect constructor).
-
+            x (PyTorch tensor): A tensor of shape `(B, C, T)`. B is a batch size, 
+                C denotes a number of channels, T is a length of signal sequence. 
         Returns:
-            y (PyTorch tensor): A tensor of shape `(B, C, Q, J, T/(2**J))`
+            y (PyTorch tensor): A tensor of shape `(B, C, Q, J, T_out)` if 
+                skip_lp=True, otherwise a tensor of shape `(B, C, Q, J, T/(2**J))`
         """
         assert self.C == x.shape[1]
-        _, bps = self.dtcwt(x)
-        ys = []
+        lp, bps = self.dtcwt(x)
+        output = []
 
         for j in range(self.dtcwt.J):
-            Wx_r = self.conv1d[j](bps[j].real)
-            Wx_i = self.conv1d[j](bps[j].imag)
-            Ux = Wx_r ** 2 + Wx_i ** 2
-            y_j, _ = self.down[j](Ux)
+            Wx_j_r = self.conv1d[j](bps[j].real)
+            Wx_j_i = self.conv1d[j](bps[j].imag)
+            UWx_j = ModulusStable.apply(Wx_j_r, Wx_j_i)
+            UWx_j, _ = self.down[j](UWx_j)
 
-            B, _, N = y_j.shape
+            B, C, N = UWx_j.shape
             # reshape from (B, C*Q, N) to (B, C, Q, N)
-            y_j = y_j.view(B, self.C, self.Q, N)
-            ys.append(y_j)
+            UWx_j = UWx_j.view(B, self.C, self.Q, N)
+            output.append(UWx_j)
+
+        return torch.stack(output, dim=3)
+
+
+class ModulusStable(torch.autograd.Function):
+    """Stable complex modulus
+
+    This class implements a modulus transform for complex numbers which is
+    stable with respect to very small inputs (z close to 0), avoiding
+    returning nans in all cases.
+
+    -------
+    Adapted from Kymatio: https://github.com/kymatio/kymatio/blob/main/kymatio/backend/torch_backend.py
```

Comment:
i don't recommend putting a URL here. "adapted from Kymatio" suffices
no need for a copyright notice, Kymatio and MuReNN are siblings 😊